File: src/main/java/org/cubexmc/railway/control/TrainControlMode.java
package org.cubexmc.railway.control;

public enum TrainControlMode {
    KINEMATIC,
    LEASHED,
    REACTIVE;

    public static TrainControlMode from(String raw, TrainControlMode fallback) {
        if (raw == null)
            return fallback;
        String v = raw.trim().toUpperCase();
        for (TrainControlMode m : values()) {
            if (m.name().equals(v))
                return m;
        }
        // also accept human-friendly strings
        if ("KINEMATIC".equalsIgnoreCase(raw))
            return KINEMATIC;
        if ("LEASHED".equalsIgnoreCase(raw))
            return LEASHED;
        if ("REACTIVE".equalsIgnoreCase(raw))
            return REACTIVE;
        return fallback;
    }
}

File: src/main/java/org/cubexmc/metro/control/TrainControlMode.kt
package org.cubexmc.metro.control

import java.util.Locale

enum class TrainControlMode {
    KINEMATIC,
    LEASHED,
    REACTIVE,
    ;

    companion object {
        @JvmStatic
        fun from(raw: String?, fallback: TrainControlMode?): TrainControlMode? {
            if (raw == null) {
                return fallback
            }
            val v = raw.trim().uppercase(Locale.getDefault())
            for (mode in values()) {
                if (mode.name == v) {
                    return mode
                }
            }
            // also accept human-friendly strings
            if ("KINEMATIC".equals(raw, ignoreCase = true)) return KINEMATIC
            if ("LEASHED".equals(raw, ignoreCase = true)) return LEASHED
            if ("REACTIVE".equals(raw, ignoreCase = true)) return REACTIVE
            return fallback
        }
    }
}
